Multiple Backends

Soft opened the pull request:
This branch adds support for alternative analyzer backends and contains an analyzer for Linux guests. For each performed system call, the analyzer finds the associated process name and the name of the system call handler responsible for handling the call.

While the branch is not currently in a state that can be automatically merged with the master, I still created this pull request to discuss if and how these changes could eventually be upstreamed. The upstream has changed quite a bit after I developed these so merging the changes could be a bit of work.

As a word of warning, I haven't done testing with Windows guests so the separated Windows backend could be currently broken. However, any breakage should be pretty easy to fix.

What do you think about these changes and do you have any opinions about integrating them?

Wenzel commented:
Hi Samuel,

thank you very much for this PR, i will review and give you my feedback as soon as I have some time to work on it.

noxdafox commented:
The PR has conflicts, could you please re-base?

noxdafox commented on main.py:
```diff
@@ -46,22 +46,21 @@ def main(args):
     if not args['--nobackend']:
-        backend = Backend(domain)
+        backend = get_backend(domain)
+        if not backend:
```
`if backend is not None`

https://www.python.org/dev/peps/pep-0008/#id51

Soft commented:
I rebased the code so that it can be merged cleanly. I've yet to implement ArgumentMap for Linux, but it shouldn't pose a problem. I also got my hands on a Windows VM and tested the separated Windows backend, which appears to be working, or at least it doesn't break on start up...

The only thing that I haven't tested are the tests in `/tests` directory, so far I've done my own testing in a more manual way. It's possible there is some breakage with those.

noxdafox commented on nitro/backends/__init__.py:
```diff
@@ -0,0 +1,21 @@
+
```
It is generally considered a bad practice to put logic in the `__init__.py` files. The core reason is `__init__.py` files are intended to describe the layout of a [package](https://docs.python.org/3/tutorial/modules.html#packages). 

In `__init__.py` files you're supposed to define and initialize the module hierarchy (helping yourself with the `__all__` variable) of your project.

It is the *last* file usually a developer goes to look into when searching a given implementation.

noxdafox commented on nitro/backends/backend.py:
```diff
@@ -0,0 +1,76 @@
+            pass
+        else:
+            try:
+                logging.debug('Processing hook {} - {}'.format(syscall.event.direction.name, hook.__name__))
```
Do not format strings when using the [`logging`](https://docs.python.org/3/howto/logging.html#optimization) library, let the library do it for you.

```
logging.debug('Processing hook %s - %s', syscall.event.direction.name, hook.__name__)
```

noxdafox commented on nitro/backends/backend.py:
```diff
@@ -0,0 +1,76 @@
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
```
Protip, you can ignore arguments with:

`def function(self, *_):`

noxdafox commented on nitro/backends/linux/arguments.py:
```diff
@@ -0,0 +1,11 @@
+
+class LinuxArgumentMap(ArgumentMap):
+    def __init__(self, event, name, process, nitro):
+        super().__init__(event, name, process, nitro)
```
You don't need to override the parent class `__init__` if both signature and implementation are the same.

noxdafox commented on nitro/backends/linux/backend.py:
```diff
@@ -0,0 +1,88 @@
+        vcpus_info = self.domain.vcpus()
+        self.nb_vcpu = len(vcpus_info[0])
+
+        self.syscall_stack = [[] for _ in range(self.nb_vcpu)]
```
I would suggest a tuple of lists here as the number of cpus is static.

noxdafox commented on nitro/backends/linux/backend.py:
```diff
@@ -0,0 +1,88 @@
+    def associate_process(self, cr3):
+        """Get Process associated with CR3"""
+        head = self.libvmi.translate_ksym2v("init_task") # get the address of swapper's task_struct
+        next = head
```
`next` is a reserved keyword in Python.

noxdafox commented on nitro/backends/windows/arguments.py:
```diff
@@ -0,0 +1,65 @@
+        try:
+            arg_type, opaque = self.CONVENTION[self.event.type][index]
+        except KeyError:
+            raise RuntimeError('Unknown covention')
```
Python 3 allows exception chaining which is quite nice when inspecting errors.

```
except KeyError as error:
    raise RuntimeError('Unknown covention') from error
```

noxdafox commented on nitro/backends/windows/backend.py:
```diff
@@ -1,73 +1,72 @@
+
+        # create on syscall stack per vcpu
+        self.syscall_stack = {}
+        for vcpu_nb in range(self.nb_vcpu):
```
Tuple here as well I guess.

noxdafox commented on nitro/backends/windows/backend.py:
```diff
@@ -1,73 +1,72 @@
-
-
-class Backend:
+GETSYMBOLS_SCRIPT = 'get_symbols.py'
```
Not in the scope of this PR.

As long as there are not good reasons for that, I would consider running this Python script within a native Python `Process` rather than using `subprocess`. There is much more overhead and less control when using `subprocess`.

Wenzel replied:
@noxdafox there are good reasons :)

`Rekall API` and more generally the `Rekall` framework is only available in Python 2.

Wenzel replied:
What do you mean by a native Python `Process` ?
How does it differ from using `subprocess` ?

noxdafox replied:
I mean [`multiprocessing.Process`](https://docs.python.org/3.6/library/multiprocessing.html#the-process-class).

Very briefly it boils down to a `fork` vs `fork`+`exec` in Unix. 

Obvious consequence is you are starting a brand new Python interpreter with `subprocess`. Also communication might suffer overhead as you are using STDIN/OUT instead of plain pipes (would need some bench test though).

Soft commented:
Thanks for the review, I tried to fix some of the issues mentioned in the latest commit. The issues in the Windows backend are not really specific to this branch as the code is mostly unaltered and just reorganized.

Wenzel commented on nitro/process.py:
```diff
@@ -1,8 +1,9 @@
+    def __init__(self, cr3, descriptor, name, pid, libvmi):
         self.cr3 = cr3
-        self.start_eproc = start_eproc
+        self.descriptor = descriptor
```
Maybe adding a comment here to state that `descriptor` is the start address of the `EPROCESS` or `task_struct` kernel structure.

Wenzel commented:
Could you give me your `libvmi.conf` ?
I would like to test your branch on `Ubuntu 16.04` guest, but i don't have the offsets.

Where did you find them ?

Soft commented:
I've been using a bunch of Gentoo systems for my testing purposes. The offsets are generated using the kernel module available in the libvmi [repository](https://github.com/KVM-VMI/libvmi/tree/master/tools/linux-offset-finder).

I've included the `libvmi.conf` here, but the offsets may obviously differ:

```
Gentoo-Nested {
    ostype      = "Linux";
    sysmap      = "/home/user/kvm-vmi/Gentoo-Clone/System.map-4.9.6-gentoo-r1";
    linux_name  = 0x620;
    linux_tasks = 0x368;
    linux_mm    = 0x3b8;
    linux_pid   = 0x468;
    linux_pgd   = 0x40;
}

win7 {
    ostype       = "Windows";
    win_tasks    = 0x188;
    win_pdbase   = 0x28;
    win_pid      = 0x180;
    win_pname    = 0x2e0;
}
```

Wenzel commented:
Thanks Samuel.

I used the offset finder and was able to retrieve them.

Here is my config now

libvmi.conf
~~~
ubuntu_1604 {
    sysmap = "/boot/System.map-4.8.0-36-generic";
    ostype      = "Linux";
    linux_name = 0x678;
    linux_tasks = 0x3c8;
    linux_mm = 0x418;
    linux_pid = 0x4c8;
    linux_pgd = 0x40;
}
~~~

`vmi-process-list` is working, so my libvmi is ready.

Wenzel commented:
I was able to test your Linux backend and get the output.

I also retested the Windows one, which is working.

What i observed is that `libvmi` is initialized twice:

~~~
LibVMI Version 0.11.0
LibVMI Driver Mode 1
--completed driver init.
--got id from name (ubuntu_1604 --> 6)
**set image_type = ubuntu_1604
--libvirt version 1003001
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "pmemaccess", "arguments": {"path": "/tmp/vmieXfRNu"}}'
--kvm: using custom patch for fast memory access
**set allocated_ram_size = 40000000, max_physical_address = 0x40000000
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
**set pae = 1
**set pse = 1
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
**set lme = 1
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
**IA-32e paging
**sanity checking cr3 = 0x000000003c6be000
--looking for config file at /home/developer/kvm-vmi/nitro/libvmi.conf
**Using config file at /home/developer/kvm-vmi/nitro/libvmi.conf
**set os_type to Linux.
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
VMI_ERROR: --requesting PA [0xfee01000] beyond max physical address [0x40000000]
VMI_ERROR: 	paddr: fee00000, length 1000, vmi->max_physical_address 40000000
VMI_ERROR: create_new_entry failed
VMI_ERROR: --requesting PA [0xfec01000] beyond max physical address [0x40000000]
VMI_ERROR: 	paddr: fec00000, length 1000, vmi->max_physical_address 40000000
VMI_ERROR: create_new_entry failed
Finding QEMU pid for domain ubuntu_1604
Detected 1 VCPUs
LibVMI Version 0.11.0
LibVMI Driver Mode 1
--completed driver init.
--got id from name (ubuntu_1604 --> 6)
**set image_type = ubuntu_1604
--libvirt version 1003001
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "pmemaccess", "arguments": {"path": "/tmp/vmiMDs7hA"}}'
--kvm: using custom patch for fast memory access
**set allocated_ram_size = 40000000, max_physical_address = 0x40000000
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
**set pae = 1
**set pse = 1
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
**set lme = 1
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
**IA-32e paging
**sanity checking cr3 = 0x0000000037d05000
--looking for config file at /home/developer/kvm-vmi/nitro/libvmi.conf
**Using config file at /home/developer/kvm-vmi/nitro/libvmi.conf
Duplicate config for ubuntu_1604 found, using most recent
**set os_type to Linux.
--qmp: virsh -c qemu:///system qemu-monitor-command ubuntu_1604 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
VMI_ERROR: --requesting PA [0xfee01000] beyond max physical address [0x40000000]
VMI_ERROR: 	paddr: fee00000, length 1000, vmi->max_physical_address 40000000
VMI_ERROR: create_new_entry failed
VMI_ERROR: --requesting PA [0xfec01000] beyond max physical address [0x40000000]
VMI_ERROR: 	paddr: fec00000, length 1000, vmi->max_physical_address 40000000
VMI_ERROR: create_new_entry failed
Start listening on VCPU 0
{'event': {'cr3': '0x37d05000',
           'direction': 'exit',
           'rax': '0x0',
           'type': 'syscall',
           'vcpu': 0},
 'full_name': None,
 'name': None,
 'process': {'name': 'compiz',
             'pid': 1330}}
~~~

Soft commented:
Hmm, yeah the double initialization is weird, I'll have to investigate it. Also, weird that the `full_name` and `name` fields in the log you included are `None`, I don't think I have seen that.

I'll try to look into these issues next.

Wenzel commented on nitro/backends/factory.py:
```diff
@@ -0,0 +1,21 @@
+
+def get_backend(domain, analyze):
+    """Return backend based on libvmi configuration. If analyze if False, returns a dummy backend that does not analyze system calls. Returns None if the backend is missing"""
+    libvmi = Libvmi(domain.name())
```
First libvmi initilization

Wenzel commented on nitro/backends/backend.py:
```diff
@@ -0,0 +1,76 @@
+    def __init__(self, domain):
+        self.domain = domain
+        self.nitro = Nitro(self.domain)
+        self.libvmi = Libvmi(domain.name())
```
second libvmi initilization.

Soft replied:
Oh right, good catch, I'll fix it right away

Wenzel commented on nitro/backends/factory.py:
```diff
@@ -0,0 +1,21 @@
+    """Return backend based on libvmi configuration. If analyze if False, returns a dummy backend that does not analyze system calls. Returns None if the backend is missing"""
+    libvmi = Libvmi(domain.name())
+    if analyze:
+        backend = BACKENDS.get(libvmi.get_ostype())
```
I would like to suggest here to fail and raise and exception, rather than returning a `None` backend.

What about this logic:

~~~Python
try:
     backend = BACKENDS.get[libvmi.get_ostype()]
     return backend
except KeyError:
    raise BackendNotFoundException('blabla')
~~~

Wenzel commented on nitro/backends/factory.py:
```diff
@@ -0,0 +1,21 @@
+        if backend is not None:
+            return backend(domain, libvmi)
+    else:
+        return Backend(domain, libvmi)
```
I will explain a bit why i choose to introduce the `analyze` backend parameter.

The whole idea was to make the Nitro object accessible to the Syscalls objects, so that they can query the registers and write new values.

To make this possible, i needed to initialize Nitro in the Backend.

But now, this makes what the "Backend" is and what it does very confusing.

We should rename it at least, like an `Engine` for example.

Wenzel replied:
It could become something like `Engine` having an `introspection_enabled` parameter.

Wenzel replied:
see my PR on your fork.
[https://github.com/Soft/nitro/pull/1](https://github.com/Soft/nitro/pull/1)

Soft replied:
I've also been thinking about what the role of `Syscall` and `Process` classes should be. Should they be just dumb data containers or should they take care of extracting the data from Nitro. At present, they of course serve only as containers for data that `Backend`s extracts.

If we wanted to move the extraction and analysis to `Syscall` and (maybe even) `Process`, it would mean they would also have to be specific to a particular Backend. I would kind of like this design, since already the one class for every backend approach seems little messy (for example, having the `descriptor` field in `Process`).

Also, my feeling is that the more data we decide to extract into these objects the more messy the one class for every backend approach becomes. For example, lets say we wanted to extract something very Windows (or Linux) specific from each process, where would we place that in the current design.

Wenzel replied:
I think this i a good idea, because otherwise, the Linux/Windows Backend class will contain too much logic.

For example to build a Process, we should just have a Linux/Windows Process class that will be as following

~~~Python

class WindowsProcess():

    def __init__(self, cr3, descriptor):
          ....
~~~

However for the `Syscall` i don't see the need for OS specific classes since the only thing that distinguishes a Linux syscall from a Windows syscall is where the parameters are stored (memory or register), which is the job of our `ArgumentMap` class.

noxdafox replied:
Can you list in order of importance the roles/responsibilities of the `Backend`? What is it? What it does?

Wenzel replied:
It's the translation/introspection layer of Nitro.

~~~
+----------+       +-------------+       +----------------------------+
| Listener +-------+  Backend    +------>+  Syscall / Process context |
+----------+       +-------------+       +----------------------------+
~~~

Role is to contain the OS logic and knowledge to rebuild the context from the raw event and informations sources that we have (`memory`, `registers`)

noxdafox replied:
From the role/responsibilities of an object, you can usually infer its name.

`Introspection`
`Interpreter`
`OSInterpreter`
`EventInterpreter`
`SymbolTranslator`

`Backend` is a terrible name as it implies it's a server. I'd expect it to be a separate process exposing a server interface.

Wenzel replied:
We already reworked the API so that the developer will use `Nitro` as the main class to interact with nitro.

and yes, `Backend` is a terrible name, we need to replace it.

Wenzel commented on nitro/backends/linux/backend.py:
```diff
@@ -0,0 +1,95 @@
+                pgd = self.libvmi.read_addr_va(mm + self.pgd_offset, 0)
+                pgd_phys_addr = self.libvmi.translate_kv2p(pgd)
+                if cr3 == pgd_phys_addr:
+                    # Eventually, I would like to look for the executable name from mm->exe_file->f_path
```
Next step would be to create a `LinuxProcess`, pass the `cr3`, `descriptor`, and the `libvmi` instance, and move as much Linux `task_struct` logic as possible there.

Soft replied:
Done, in the present for we only extract name and pid in the LinuxProcess class

Soft commented:
What do you think are the main missing features and problems that should be fixed before the code could in principle be upstreamed? The two things that come to mind are improved argument handling and integrating with the included tests.

Wenzel commented:
We still miss the Linux ArgumentMap.
I proposed you an implementation in a PR.

This should be trivial since the convention is very strict (always 6 registers).

And having some tests on certain syscalls would not be too much to validate this new backend system :)

Soft commented:
Thanks for the ArgumentMap implementation, I merged it with some fixes.

Wenzel commented:
That's a good catch !

while rereading this https://lwn.net/Articles/604515/
i noticed that sentence 
> Moving outward, sys_call_table is accessed from the ia32_sysenter_target entry point of arch/x86/kernel/entry_32.S. However, here the SAVE_ALL macro actually pushes a different set of registers (EBX/ECX/EDX/ESI/EDI/EBP rather than RDI/RSI/RDX/R10/R8/R9) onto the stack, reflecting the different syscall ABI convention for this platform. 

I will fix my issue description right away.

Wenzel commented:
Status update: the last thing we miss is some testing to confirm that our Linux ArgumentMap is valid.
@Soft do you plan to write the tests ?

Soft commented:
Yeah I was just planning on how to implement the testing for this.

I see that you are using Packer for creating the test image. I was thinking of creating a Packer configuration for our Linux images with some additional test binaries. Then I'll write something like `test_windows.py` for testing out the system.

Wenzel commented:
i have  a branch where i refactored the packer templates.
I added a template for Ubuntu there, you should pick it up and use it.
check this: https://github.com/KVM-VMI/nitro/tree/vm_templates/tests/vm_templates

For this PR, just push a test_file.py that i can run.
I will merge the `vm_templates` branch (with ubuntu packer templates, refactored) later on.

Soft commented:
Thanks I'll use that as a starting point

Wenzel commented:
Actually i just realized that we need to rethink the test framework.

It has been designed for Windows only, and the cdrom which is inserted to run the test payload is Windows based (autorun, bat, powershell)

We need to configure the ubuntu template to automatically execute a specific file when a CDROM is inserted, and add the cdrom preparation logic in `vm_helper.py`.
Is there an equivalent of `Autorun.inf` in Linux ?

Soft commented:
I think the cleanest way would be to setup an ssh access into the system and use it to transfer the test binaries and execute them.

Soft commented:
Or maybe place some simple http daemon that would accept payloads and execute them. Doesn't See have something like this?

Wenzel commented:
We already have the CDROM logic to inject a test payload in Windows.
I don't want to have to maintain 2 separate testing API for Windows and Linux.

The idea here would be to install the GNOME or KDE environement, since they support features like autorun, and configure them.

Soft commented:
I don't think we necessarily need a certain DE for autorun, a simple udev rule would likely be enough to execute a custom script when CDROM is attached.

Wenzel commented:
Hi @Soft ,
i just merged some small PR, and i added the vm templates too.
you should rebase your work on top of it now ;)

Soft commented:
Oh thanks, I had already cherry picked the vm_templates changes from your branch but I'll try to do a rebase later, it is getting quite complicated with the diverging histories.

I am now in the middle of implementing the testing stuff for Linux. My plan so far is to alter CDROM.py to  write a shell script that runs the desired commands, modify the packer template for Ubuntu to add an udev rule and some additional scripts for automounting and executing the script from the generated CDs, and modify vmtest_helper.py to work with Linux. For the last one, I'll figure out some way to inform the host that the test scripts finished executing. After this, I can finally write some actual test cases.

Soft commented:
I pushed few changes to the ubuntu template. Now the ubuntu test systems automatically mount CDROMs to /mnt/cdrom and execute special autoexec.sh scripts from them if they exist.

Wenzel commented:
Ok, next step is to write some tests to verify that we are mapping the arguments right.

could you rebase your branch on master ?
I see some of my commits in your PR. 
like `check KVM_NITRO_ATTACH_VCPUS ioctl return value`, which is already integrated into master.

Soft commented:
Yeah I'll get to that but first I'll have to update CDROM generation and vmhelper. I haven't done a full rebase yet, instead I cherry picked some of the missing commits such as the vm_templates ones.

Wenzel commented:
Hi @Soft 

i merged some PR, and there is some ongoing changes that i want to inform you about.

First, this commit : https://github.com/KVM-VMI/nitro/commit/6e31f54f12c856ab11e3aceaeed9fff21e3a2041
which reuses the `Syscall` object at exit instead or creating a new one.

Then some changes : https://github.com/KVM-VMI/nitro/commit/09e4291133a198680bae8778c8f8d6c0a04286c6
in `win_types`, but you should not really care.

And now i need to reintroduce the `Backend` object as a callback parameter.
PR: https://github.com/KVM-VMI/nitro/pull/25

Reason for that is that the `Backend` keeps track of the running processes, and a callback might want to access this process list and dig into a Process object.

Finally some refactoring has been done in the Process class, a PR is waiting : 
https://github.com/KVM-VMI/nitro/pull/24

Thanks !

Soft commented:
Okay, thanks for the heads up. I'll try to rebase this branch as soon as possible, but I would like to complete my integration testing work before, or at least bring them a bit closer to completion.

Today, I've been spending time moving my testing environment to a serparate server, hopefully I'll get back to doing actual development soon.

Soft commented:
Since rebases seem to get really painful with the growing history of this branch, I just merged the changes from the upstream master to my branch, hopefully I didn't mess up anything.

I've been writing rudimentary tests for the Linux backend as most of the required infrastructure seems to be in place. I have tested the general attach a CDROM → tests get executed → sshd gets shutdown logic without Nitro and everything works well. However, when I tried to automate this and observe the effects using Nitro things seemed to break. I've been spending time debugging the issue and it seems that the slowdown caused by Nitro leads to systemd-udev giving up on processing the CDROM attached event before the included test script is executed. I modified my Ubuntu test image to dump bunch of debug logs over serial to see this:

```
[  145.047973] systemd-udevd[262]: seq 2001 '/devices/pci0000:00/0000:00:01.1/ata1/host0/target0:0:1/0:0:1:0/block/sr0' is taking a long time
[  145.436841] systemd-udevd[262]: seq 2001 '/devices/pci0000:00/0000:00:01.1/ata1/host0/target0:0:1/0:0:1:0/block/sr0' is taking a long time
```

...after some time...

```
[  519.251670] systemd-udevd[262]: worker [677] terminated by signal 9 (Killed)
```
I solved this by increasing the event-timeout parameter for udev and changed the system so that it doesn't run the test script directly from udev but instead simply uses udev to trigger a systemd service that then, in turn, runs the tests. This seems to work since unlike udev's RUN directives, systemd doesn't care if commands execute slowly. I think this is in general a more sensible design.

I also noticed that the Linux backend can produce errors. These seem to appear mainly during bootup and shutdown. The cause for these was libvmi's internal caching. One can remedy this by flushing the caches in the backend when processing events but this will lead to an even greater performance penalty. Additionally, I changed the event production pipeline slightly so that it is possible for backends to fail at analyzing events. Of course, trying to do analysis with potentially invalid caches is probably not the best idea. However, as an additional layer of robustness this might be useful.

All and all, the Linux backend testing code seems to be in a good shape. I'll add more actual tests cases next.

I hope we will get this branch merged soon. I would like to move on doing the documentation and unit testing for the project.

---

As a side note that I should probably move to the issues tracker over at the kvm project, there seems to be a bug in the code where detaching Nitro in the wrong way causes soft lockups in guest kernels. You can even see this if you monitor the CPU usage charts in virt-manager, when these incidents happen the guest CPU usage will go through the roof. I have not yet investigated the possible causes for this.

Soft commented:
As a quick update, we now have integration tets in place and, for example, argument access seems to be working okay based on those results.

Also I've been doing some ground work on getting tools in place for easy definition of in-VM data structures. This work can be found on the ``foreign`` branch as I did not want to push anything here that could make merging this take even more time. My plan is to provide a set of classes and glue for defining data structure layouts, loading them from memory and updating changes back. It is still very much experimental and some features are missing. Here's some relevant [code](https://github.com/Soft/nitro/commit/4ccf5cec6348a67a8e4932c90c53c67c3c25abd0).

aghamir commented:
Hi @Soft,
I used your `linux` branch for nitro. However when I run your code and halt it by `CTRL+C`, I cannot use GVM anymore.
Maybe this causes by semaphore pause at `nitro KVM` but I'm not sure about it.
Do you have the same issue? I didn't check it in upstream nitro but maybe @Wenzel knows the solution.

Soft commented:
@aghamir thanks for your report.

Could you maybe describe the error in a little more detail. What does the KVM report after you try to use it after using Nitro? How does the error manifest itself?

I know there are some bugs that seem to be triggered when ``CTRL+C``'ing out of Nitro, mainly, that the monitored guests seem to freeze in some way, or at least their CPU usage jumps while the system grinds to a halt. From the hosts side I can still shutdown the malfunctioning guest and start it again...

This is of course something that I would like to fix and asume that problem is in the KVM patches. However, I stil haven't had time to look too much into it. Any hints about the possible cause would obviously be greatly appreciated.

aghamir commented:
Thank you for your fast response. You exactly mention my error.
`I know there are some bugs that seem to be triggered when CTRL+C'ing out of Nitro, mainly, that the monitored guests seem to freeze in some way, or at least their CPU usage jumps while the system grinds to a halt. From the hosts side I can still shutdown the malfunctioning guest and start it again...`
Freezing os is the exactly error I faced. I didn't log my KVM to say what exactly happened after halting nitro. As far as I know, kvm uses `semaphore` in `nitro_ioctl_get_event` to report to userland. It also uses `completion` to pause vm for its report. However its `completion` has timeout. So I am suspecting to `semaphore` in kvm. 

BTW, is there any issue tracking this freeze error?

Soft commented:
@aghamir I created an issue regarding the problem

Wenzel commented:
HI @Soft ,
sorry for the late review.

# Udev timeouts
I saw your issues with udev timeout caused by Nitro, and i have the same on Windows.
Sometimes the CDRom execution logic isn't executed, and a windows appears to say it's going to be processed soon, but it isn't.

I haven't found a fix for this (it was only appearing when i was on nested virtualization, so this + the nitro overhead was really disturbing some Windows internal timeouts).

But it's good that searched a workaround with systemd.

# Libvmi caching issues
I had the same, and yes `libvmi` has an internal cache that needs to be flushed.
However, as i discovered, there was no way to flush the internal page cache, so i opened an issue about it:
https://github.com/libvmi/libvmi/issues/464

For the moment, i forked libvmi, and changed it so that when you flush the `v2pcache`, it flushes also the internal memcache.

That's the reason i'm flushing every possible cache when a new event is being processed in the Backend:
https://github.com/KVM-VMI/nitro/blob/master/nitro/backend.py#L126

And the first call to `v2pcacheflush` will also flush the internal one if you use our `libvmi`.

# merge
About the merge, it would have been better to rebase your work on top of the master branch.
That's how we should do, and you could have used that to squash some of your commits and rephrase some of them (i don't know if you are used to that process ?)

# documentation and tests
I advise you to create 2 different PR for the documentation and the tests, because this one is already very big and painful to read.

Thanks !

Soft commented:
Thanks for the reply, that comment about internal page cache seems really interesting, that might fix some of the problems I am seeing.

As it stands now, this branch doesn't have anything special in terms of documentation. I have Sphinx running locally and I've been trying to write up something but I still haven't pushed it anywhere. I agree that I should place those in a separate branch. My plan was to create a separate PR for docs and unit tests, as we discussed last time.

And yes, rebasing would maybe have been cleaner, I already did that once with this branch when the history was a bit shorter. However, rebasing can get a bit complicated when the history of the branch gets long.

aghamir commented:
Hi @Soft ,
I found the freezing issue. It cause by closing fd after finishing nitro. Just comment `self.kvm_io.close()` in `listener.py`.

Soft commented:
@aghamir Thank you so much for your work on this issue. That seems really promising. I just wonder if that fix could cause some other side effects. Is it okay to just not call close on kvm's file descriptor. I'll look into the matter tomorrow and do some testing.

Soft commented:
I updated this branch again so that it does not have conflicts. I also updated Windows debug tests to work with the modified testing infrastructure.

Soft commented:
Hi, I think this branch is now in a mergeable state. Could this be merged as we discussed last time?

Wenzel commented:
Hi @Soft ,

i added a few comments that needs to be resolved since last time.
can you take a look at them ?

Soft commented:
@Wenzel: Where are the comments you are referring to?

Wenzel commented:
@Soft sorry, i thought i just had to put some comments to get a review here.
But i had to explicitely ask for a review, like i did just now

Soft commented:
I have fixed most of the things that came up in the review

Soft commented:
What is the status on this one?

Soft commented:
If there are any remaining issues could we just make them into separate issues like we discussed in the last meeting and get this merged. Or is there something truly blocking?

Wenzel commented:
Hi @soft,

I saw that you fixed some stuff, but i didn't had the chance to review them.

I will be available for this review on thursday.

Thanks :)

Wenzel commented:
Hi @Soft, i tested and approved your changes.
Also the linux backend is working.

Let's merge this PR ;)
